packages/aws-cdk-lib/core/lib/nodejs-aspect.ts

```diff
@@ -0,0 +1,138 @@
+/* eslint-disable brace-style */
+import { Construct, IConstruct } from 'constructs';
+import { IAspect, Aspects } from './aspect';
+import * as cdk from '../../../aws-cdk-lib';
+import * as lambda from '../../aws-lambda';
+import { ReceiptRuleSet } from '../../aws-ses/lib/receipt-rule-set';
+import { Provider } from '../../custom-resources/lib';
+
+/**
+ * Runtime aspec
+ */
+abstract class RuntimeAspectsBase implements IAspect {
+
+  /**
+   * The string key for the runtime
+   */
+  public readonly targetRuntime: string;
+
+  constructor(runtime: string) {
+    this.targetRuntime = runtime;
+  }
+
+  public visit(construct: IConstruct): void {
+
+    //To handle custom providers
+    if (construct instanceof cdk.CustomResourceProviderBase) {
+      this.handleCustomResourceProvider(construct);
+    }
+
+    //To handle providers
+    else if (construct instanceof Provider) {
+      this.handleProvider(construct);
+    }
+
+    //To handle single Function case
+    else if (construct instanceof ReceiptRuleSet) {
+      this.handleReceiptRuleSet(construct);
+    }
+  }
+
+  private handleCustomResourceProvider(provider: cdk.CustomResourceProviderBase): void {
+    const node = provider as cdk.CustomResourceProviderBase;
+    // only replace for nodejs case
+    if (this.isValidRuntime(node.runtime)) {
+      const targetnode = node.node.children.find((child) => child instanceof cdk.CfnResource &&
+      child.cfnResourceType === 'AWS::Lambda::Function') as cdk.CfnResource;
+      targetnode.addPropertyOverride('Runtime', 'nodejs20.x');
+    }
+  }
+
+  private handleProvider(provider: Provider): void {
+    const functionNodes = provider.node.children.filter((child) => child instanceof lambda.Function);
+    for ( var node of functionNodes) {
+      const targetNode = node.node.children.find((child) => child instanceof cdk.CfnResource && child.cfnResourceType === 'AWS::Lambda::Function') as cdk.CfnResource;
```

Comment:
A few things about this logic:
1. You have this same thing happening in three different spots. This is something that can be factored out.
2. You're visiting child nodes that will be visited anyways. Let me explain. All custom resource providers are just Lambda functions. This means they all extend `CfnResource` and have a `cfnResourceType` of `AWS::Lambda::Function`. You should only need to just have logic that checks for an instance of `CfnResource` and a `cfnResourceType` of `AWS::Lambda::Function`. You don't need all these branches.